utils/augmentation.py

```diff
@@ -355,13 +356,19 @@ def __call__(self, sample):
 
 class ToTensorTarget(object):
     """Convert ndarrays in sample to Tensors."""
+
+    def __init__(self, num_classes):
+        self.num_classes = num_classes
+
     def __call__(self, sample):
         sat_img = np.nan_to_num(sample['sat_img'], copy=False)
         sat_img = np.float32(np.transpose(sat_img, (2, 0, 1)))
         sat_img = torch.from_numpy(sat_img)
 
         map_img = None
-        if 'map_img' in sample.keys():  # This can also be used in inference.
-            map_img = np.int64(sample['map_img'])
-            map_img = torch.from_numpy(map_img)
+        if 'map_img' in sample.keys():
+            if sample['map_img'] is not None:  # This can also be used in inference.
+                map_img = np.int64(sample['map_img'])
+                map_img[map_img > self.num_classes] = 0
+                map_img = torch.from_numpy(map_img)
```

Comment:
doesn't this affect the values at 255 that could be passed as "ignore_index" ? I know you're not a fan or ignore_index, @victorlazio109...

Reply:
Yes! for the following reasons:
- we allow un-wanted values to pass through at data creation which introduces inconsistencies. At the point of feeding data into the model, I want to be sure my values are correct in the right order. Passing labels as [0, 1, 2, 3] vs [1, 2, 3, 255] (255 is greater than the num of categorical classes we define or are looking to detect)
- ignore_index is only accounted for in the loss function that supports that flag. There are losses that expect only correct number of categorical classes there is no support for ignore_index
- I think there is simplicity in passing only the required num of classes
This might not be the best implementation, however, if we must pass an --ignore_index flag to our loss function we can certainly isolate it and it is best practice to have the class ignored within the range of the classes you are trying to detect.

Reply:
Get it. It's true, not all losses have ignore_index implemented and I see how that could cause problems. Now, we'd need to see what would happen to parts in the code where where ignore_index is used. For example, in images_to_samples.py, we overwrite label values with ignore_index value if pixel corresponds to nodata value in image:
https://github.com/NRCan/geo-deep-learning/blob/develop/images_to_samples.py#L439

If there are large parts of the image that nodata values, then it seems logical to me that we don't want the model to learn on prediction in that area, thus that overwritting... I'd like to talk it over with you and other teammembers. Maybe at our meeting tomorrow am.